internal/controller/siteconfig_helper.go

```diff
@@ -0,0 +1,306 @@
+	if !json.Valid([]byte(networkAnnotation)) {
+		return installConfigOverrides, fmt.Errorf("invalid json conversion of network type")
+	}
+
+	switch installConfigOverrides {
+	case "":
+		return networkAnnotation, nil
+
+	default:
+		if !json.Valid([]byte(installConfigOverrides)) {
+			return "", fmt.Errorf("invalid json parameter set at installConfigOverride")
+		}
+
+		var installConfigOverridesMap map[string]interface{}
+		err := json.Unmarshal([]byte(installConfigOverrides), &installConfigOverridesMap)
+		if err != nil {
+			return "", fmt.Errorf("failed to unmarshal installConfigOverrides data: %v", installConfigOverrides)
+		}
+
+		if _, found := installConfigOverridesMap[commonKey]; found {
+			networkMergedJson, err := mergeJsonCommonKey(networkAnnotation, installConfigOverrides, commonKey)
+			if err != nil {
+				return "", fmt.Errorf("failed to merge installConfigOverrides objects, error: %v", err)
+			}
+			return networkMergedJson, nil
+		}
+
+		trimmedConfigOverrides := strings.TrimPrefix(installConfigOverrides, "{")
+		trimmedNetworkType := strings.TrimSuffix(networkAnnotation, "}")
+		finalJson := trimmedNetworkType + "," + trimmedConfigOverrides
+		if !json.Valid([]byte(finalJson)) {
+			return "", fmt.Errorf("failed to marshal annotation for installConfigOverrides, error: %v", err)
+		}
+		return finalJson, nil
+
+	}
+}
+
+// buildSiteData returns a Site object that is consumed for rendering templates
+func buildSiteData(siteConfig *v1alpha1.SiteConfig, nodeId int) (data interface{}, err error) {
+
+	// Prepare specialVars
+	var currentNode v1alpha1.NodeSpec
+	if nodeId >= 0 {
+		currentNode = siteConfig.Spec.Nodes[nodeId]
+	}
+
+	installConfigOverrides, err := getInstallConfigOverrides(siteConfig)
+	if err != nil {
+		installConfigOverrides = ""
+	}
+
+	// Determine the number of control-plane and worker agents
+	controlPlaneAgents := 0
+	workerAgents := 0
+	for _, node := range siteConfig.Spec.Nodes {
+		switch node.Role {
+		case "master":
+			controlPlaneAgents++
+		case "worker":
+			workerAgents++
+		}
+	}
+
+	data = SiteData{
+		Site: siteConfig.Spec,
+		SpecialVars: SpecialVars{
+			CurrentNode:            currentNode,
+			InstallConfigOverrides: installConfigOverrides,
+			ControlPlaneAgents:     controlPlaneAgents,
+			WorkerAgents:           workerAgents,
+		},
+	}
+
+	return
+}
+
+// suppressRenderingManifest function returns true if the manifest-rendering should be suppressed
+func suppressRenderingManifest(kind string, suppressedManifests []string) bool {
+	if kind == "" || len(suppressedManifests) == 0 {
+		return false
+	}
+
+	for _, manifest := range suppressedManifests {
+		if manifest == kind {
+			return true
+		}
+	}
+	return false
+}
+
+// mergeJsonCommonKey merge 2 json in common key and return string
+func mergeJsonCommonKey(mergeWith, mergeTo, key string) (string, error) {
+
+	var (
+		networkAnnotation      map[string]interface{}
+		installConfigOverrides map[string]interface{}
```

Comment:
You've created awfully specific variable names here for something that feels like it should be generic.
Also I think you can just json unmarshal the maps on top of each other and it should work.

Reply:
lol!!! refactored!